content/blog/2019-11-01_react-typescript-webpack-2.md

```diff
@@ -0,0 +1,193 @@
+The script in that case is `"build:prod": "cross-env NODE_ENV=production webpack"`.
+Again, if you do not need Windows support, simply remove the `cross-env` from the script.
+
+Now there are two ways to proceed:
+
+1. split the configuration in multiple files and merge them with something like [webpack-merge](https://www.npmjs.com/package/webpack-merge)
+2. keep everything in one file and use JavaScript to toggle things
+
+I prefer the second approach as I find it more readable and doesn't require additional dependencies. You can judge for yourself
+by looking at the [final result](https://github.com/Keats/webpack-react-typescript/blob/master/webpack.config.js).
+
+As mentioned before, Webpack has been making things easier. With the introduction of [mode](https://webpack.js.org/configuration/mode/), it is easy
+to get a production-ready bundle:
+
+```diff
+diff --git a/webpack.config.js b/webpack.config.js
+index 53c39f2..b4d6e11 100644
+--- a/webpack.config.js
++++ b/webpack.config.js
+@@ -4,9 +4,11 @@ const HtmlWebpackPlugin = require("html-webpack-plugin");
+ const ForkTsCheckerWebpackPlugin = require("fork-ts-checker-webpack-plugin");
+ const { CleanWebpackPlugin } = require("clean-webpack-plugin");
+ 
++const isProd = process.env.NODE_ENV === "production";
++
+ module.exports = {
+   context: __dirname,
+-  mode: "development",
++  mode: isProd ? "production" : "development",
+   entry: {
+     app: "./src/index.tsx"
+   },
+```
+
+Just changing the mode will get us a minified output: you can check it by running `build:prod` yourself. 
+
+Looking at the minified code, we can see that the sourcemap is embedded in the file still. 
+The [best option for production environments](https://webpack.js.org/configuration/devtool/#production)
+is `source-map` which we can set conditionally on the `isProd` variable: `devtool: isProd ? "source-map" : "eval-source-map",`.
+Running `build:prod` again will now properly create a separate file ending in `.js.map`.
+Remember that the sourcemap files should not be accessible by anyone other than you and your bug reporting app.
+
+Lastly, the HMR from Webpack is injecting some code in the bundle that we do not want in our production bundle. We can disable it easily:
+
+```diff
+diff --git a/webpack.config.js b/webpack.config.js
+index af5fd2f..63272fd 100644
+--- a/webpack.config.js
++++ b/webpack.config.js
+@@ -58,6 +58,6 @@ module.exports = {
+       async: false
+     }),
+     new CleanWebpackPlugin(),
+-    new webpack.HotModuleReplacementPlugin()
+-  ]
++    isProd ? false : new webpack.HotModuleReplacementPlugin()
++  ].filter(Boolean)
+ };
+```
+
+
+The commit is <https://github.com/Keats/webpack-react-typescript/commit/4527184f2d7a49447451774b1bf5cc8d4cb3c1e5>.
+
+## Splitting chunks
+In our current situation, we generate a single file for the whole app: it clocks in at 132KB before gzip.
```

Comment:
unrelated: recently remembered about the lzma2 (xz) compression format. Wondering if that could be a good alternative to gzip on browsers. Wondering how many browsers support it and could auto-unpack lz files like they do gzip.

Reply:
il y aussi Brotli mais niveau support c'est pas encore 100%. Je connaissais pas lzma2